File: model/python/injection_scheduler.py
"""Injection scheduler: architecture modes A/B/C/D and tap+DTC decode.

Contract: MODEL_SPEC.md sections 5, 7, 8 [EXACT].

    u_INJ_ideal[k] = wrap01(z0 - x_nominal[k])

Modes (feedback always A_FB = Q_FB(A_ideal)):
    A: R_INJ = Q_INJ(G * u_INJ_ideal) mod G, independent instance of the same
       quantizer type (independent error state; dither from 'dither_inj').
    B: same as A but both sides are DSM-type quantizers with their own
       state/stream (spec: correct on average, wrong cycle-by-cycle reverse;
       "not recommended").  The code path is identical to A — the mode label
       documents intent; cfg.quantizer should be 'ef1'/'mash11'.
    C: shared high-resolution master state P[k] = A_ideal[k]; u_INJ_ideal is
       derived from it, then quantized by an independent instance.
    D (default): quantize once + modular reverse:
       R_INJ = (R_zero - R_FB) mod 256  ->  e_pair_digital == 0 exactly.

Independent DSM state (spec section 7, modes A/B/C): when the quantizer is a
stateful DSM type the injection-side instance is seeded with a DISTINCT
deterministic initial state drawn from the 'dsm_inj' PRNG stream (section 12,
offset 10):
    ef1     -> e0 = one uniform draw
    mash11  -> acc1, acc2 = two uniform draws (c2_prev stays 0)
    mash111 -> acc1, acc2, acc3 = three uniform draws (in that order;
               c2_prev, c3_prev, c3_prev2 stay 0)
nearest/floor/truncate are stateless and unaffected.

Actuator mode 'dsm_only' (spec section 7.1): the fractional injection
actuator is absent — R_INJ = 0, tap j = 0, c_INJ = 0 on every cycle
(regardless of arch_mode; no 'dsm_inj' draws are consumed).

Actuator mode 'qnc' (spec section 7.2): the injection side is ALWAYS the
modular reverse of the cancellation-DTC code, R_INJ = (R_zero - R_FB) mod G,
regardless of arch_mode (no 'dsm_inj' draws are consumed); tap/DTC decode
proceeds as usual.

Mappings (input digital code R_INJ; target u_target = R_INJ/G):
    naive      : j = floor(R/32), c = R mod 32   (lower half of DTC range only)
    nearest    : argmin over j in 0..7, c in 0..63 of
                 |wrapCycles(u_target - (j/8 + c/256))|, tie-break smaller c
                 then smaller j (candidates enumerated c-major so argmin's
                 first-minimum rule implements the tie-break exactly)
    calibrated : argmin using actual tap + DTC + route models.

[Resolved ambiguity: spec section 8 allows the mapping input to be either the
digital code R_INJ or a real target; this model always uses the digital
command target u_target = R_INJ/G so the R_INJ produced by the architecture
mode remains the single command word.  e_INJ_abs is still measured against
the real u_INJ_ideal.]
"""

import numpy as np

from .feedback_scheduler import make_quantizer
from .phase_math import wrap01_arr, wrap_cycles_arr


def u_inj_ideal(x_nominal: np.ndarray, z0: float) -> np.ndarray:
    """Ideal normalized injection command: wrap01(z0 - x_nominal)."""
    return wrap01_arr(z0 - x_nominal)


def _candidates(cfg, dtc_inj, tap_tbl, calibrated: bool):
    """Candidate (j, c) list in c-major order with phase values (cycles)."""
    n_dtc = 1 << cfg.b_dtc
    cs, js, us = [], [], []
    for c in range(n_dtc):
        for j in range(cfg.n_tap):
            cs.append(c)
            js.append(j)
            if calibrated:
                us.append(tap_tbl[j] + float(dtc_inj.delay_cycles(c)) + cfg.route_inj_cycles)
            else:
                us.append(j / cfg.n_tap + c / cfg.g)
    return (np.array(js, dtype=np.int64), np.array(cs, dtype=np.int64),
            np.array(us, dtype=np.float64))


def run_injection(cfg, x_nominal: np.ndarray, r_fb: np.ndarray,
                  dtc_inj, tap_tbl, dither_stream=None,
                  dsm_stream=None) -> dict:
    """Produce injection command stream (computed-domain, pre-latency)."""
    g = cfg.g
    n = len(x_nominal)
    tap_step = g // cfg.n_tap  # 32 LSB

    u_ideal = u_inj_ideal(x_nominal, cfg.z0_cycles)

    # --- actuator mode 'dsm_only' (spec section 7.1): no fractional actuator
    if cfg.actuator_mode == "dsm_only":
        zeros = np.zeros(n, dtype=np.int64)
        u_inj_analog = np.full(
            n, tap_tbl[0] + float(dtc_inj.delay_cycles(0)) + cfg.route_inj_cycles)
        return {
            "u_INJ_ideal": u_ideal,
            "R_INJ": zeros,
            "j_INJ": zeros.copy(),
            "c_INJ": zeros.copy(),
            "u_INJ_digital": np.zeros(n, dtype=np.float64),
            "u_INJ_analog": u_inj_analog,
        }

    # --- R_INJ per architecture mode ---
    # ('qnc', spec section 7.2: always the modular reverse of the
    # cancellation-DTC code, regardless of arch_mode)
    r_inj = np.empty(n, dtype=np.int64)
    if cfg.arch_mode == "D" or cfg.actuator_mode == "qnc":
        r_inj[:] = (cfg.r_zero - r_fb) % g
    else:  # A, B, C: independent quantizer instance on G * u_INJ_ideal
        q = make_quantizer(cfg.quantizer)
        # distinct deterministic DSM state per spec section 7 ('dsm_inj' stream)
        if dsm_stream is not None:
            if cfg.quantizer == "ef1":
                q.e = dsm_stream.next()
            elif cfg.quantizer == "mash11":
                q.acc1 = dsm_stream.next()
                q.acc2 = dsm_stream.next()
            elif cfg.quantizer == "mash111":
                q.acc1 = dsm_stream.next()
                q.acc2 = dsm_stream.next()
                q.acc3 = dsm_stream.next()
        use_dither = cfg.dither_amp_lsb > 0.0 and dither_stream is not None
        for k in range(n):
            u = g * float(u_ideal[k])
            if use_dither:
                u += cfg.dither_amp_lsb * (dither_stream.next() + dither_stream.next() - 1.0)
            r_inj[k] = q.quantize(u) % g

    # --- tap + DTC decode ---
    j_inj = np.empty(n, dtype=np.int64)
    c_inj = np.empty(n, dtype=np.int64)
    if cfg.inj_mapping == "naive":
        j_inj[:] = r_inj // tap_step
        c_inj[:] = r_inj % tap_step
    else:
        calibrated = cfg.inj_mapping == "calibrated"
        js, cs, us = _candidates(cfg, dtc_inj, tap_tbl, calibrated)
        for k in range(n):
            target = r_inj[k] / g
            err = np.abs(wrap_cycles_arr(target - us))
            i = int(err.argmin())  # first minimum -> smaller c, then smaller j
            j_inj[k] = js[i]
            c_inj[k] = cs[i]

    u_inj_digital = ((tap_step * j_inj + c_inj) % g) / g
    u_inj_analog = (tap_tbl[j_inj] + dtc_inj.delay_cycles(c_inj)
                    + cfg.route_inj_cycles)

    return {
        "u_INJ_ideal": u_ideal,
        "R_INJ": r_inj,
        "j_INJ": j_inj,
        "c_INJ": c_inj,
        "u_INJ_digital": u_inj_digital,
        "u_INJ_analog": u_inj_analog,
    }


def e_pair(u_fb: np.ndarray, u_inj: np.ndarray, r_zero: int, g: int) -> np.ndarray:
    """Pairwise reverse error (digital or analog layer, spec section 7):

    e_pair = wrapCycles(u_FB_actual + u_INJ_actual - u_zero_offset)
    u_zero_offset = R_zero / G
    """
    return wrap_cycles_arr(u_fb + u_inj - r_zero / g)

File: web/src/model/injectionScheduler.ts
/**
 * Injection scheduler: architecture modes A/B/C/D and tap+DTC decode.
 *
 * Contract: MODEL_SPEC.md sections 5, 7, 8 [EXACT].
 * Mirror of model/python/injection_scheduler.py.
 *
 *     u_INJ_ideal[k] = wrap01(z0 - x_nominal[k])
 *
 * Modes (feedback always A_FB = Q_FB(A_ideal)):
 *     A: R_INJ = Q_INJ(G * u_INJ_ideal) mod G, independent instance of the
 *        same quantizer type (independent error state; dither 'dither_inj').
 *     B: same as A but both sides are DSM-type quantizers with their own
 *        state/stream ("not recommended"; code path identical to A — the
 *        mode label documents intent; cfg.quantizer should be ef1/mash11).
 *     C: shared high-resolution master state P[k] = A_ideal[k]; u_INJ_ideal
 *        derived from it, then quantized by an independent instance.
 *     D (default): quantize once + modular reverse:
 *        R_INJ = (R_zero - R_FB) mod 256  ->  e_pair_digital == 0 exactly.
 *
 * Independent DSM state (spec section 7, modes A/B/C): when the quantizer is
 * a stateful DSM type the injection-side instance is seeded with a DISTINCT
 * deterministic initial state drawn from the 'dsm_inj' PRNG stream
 * (section 12, offset 10):
 *     ef1     -> e0 = one uniform draw
 *     mash11  -> acc1, acc2 = two uniform draws (c2_prev stays 0)
 *     mash111 -> acc1, acc2, acc3 = three uniform draws (in that order;
 *                c2_prev, c3_prev, c3_prev2 stay 0)
 * nearest/floor/truncate are stateless and unaffected.
 *
 * Actuator mode 'dsm_only' (spec section 7.1): the fractional injection
 * actuator is absent — R_INJ = 0, tap j = 0, c_INJ = 0 on every cycle
 * (regardless of arch_mode; no 'dsm_inj' draws are consumed).
 *
 * Actuator mode 'qnc' (spec section 7.2): the injection side is ALWAYS the
 * modular reverse of the cancellation-DTC code,
 * R_INJ = (R_zero - R_FB) mod G, regardless of arch_mode (no 'dsm_inj'
 * draws are consumed); tap/DTC decode proceeds as usual.
 *
 * Mappings (input digital code R_INJ; target u_target = R_INJ/G):
 *     naive      : j = floor(R/32), c = R mod 32   (lower half of DTC range)
 *     nearest    : argmin over j in 0..7, c in 0..63 of
 *                  |wrapCycles(u_target - (j/8 + c/256))|, tie-break smaller
 *                  c then smaller j (candidates enumerated c-major so the
 *                  first-minimum rule implements the tie-break exactly)
 *     calibrated : argmin using actual tap + DTC + route models.
 *
 * [Resolved ambiguity: spec section 8 allows the mapping input to be either
 * the digital code R_INJ or a real target; this model always uses the digital
 * command target u_target = R_INJ/G so the R_INJ produced by the architecture
 * mode remains the single command word.  e_INJ_abs is still measured against
 * the real u_INJ_ideal.]
 */

import type { SimConfig } from './config';
import { configG } from './config';
import type { DTCModel } from './dtcModel';
import { pymod, wrap01, wrapCycles, wrapCyclesArr } from './phaseMath';
import { ErrorFeedbackFirstOrder, Mash11, Mash111, makeQuantizer } from './quantizers';
import type { Mulberry32 } from './rng';

export interface InjectionResult {
  u_INJ_ideal: Float64Array;
  R_INJ: Float64Array;
  j_INJ: Float64Array;
  c_INJ: Float64Array;
  u_INJ_digital: Float64Array;
  u_INJ_analog: Float64Array;
}

/** Ideal normalized injection command: wrap01(z0 - x_nominal). */
export function uInjIdeal(xNominal: ArrayLike<number>, z0: number): Float64Array {
  const out = new Float64Array(xNominal.length);
  for (let k = 0; k < xNominal.length; k++) {
    out[k] = wrap01(z0 - xNominal[k]);
  }
  return out;
}

/** Candidate (j, c) list in c-major order with phase values (cycles). */
function candidates(
  cfg: SimConfig,
  dtcInj: DTCModel,
  tapTbl: ArrayLike<number>,
  calibrated: boolean,
): { js: Int32Array; cs: Int32Array; us: Float64Array } {
  const nDtc = 1 << cfg.b_dtc;
  const g = configG(cfg);
  const total = nDtc * cfg.n_tap;
  const js = new Int32Array(total);
  const cs = new Int32Array(total);
  const us = new Float64Array(total);
  let i = 0;
  for (let c = 0; c < nDtc; c++) {
    for (let j = 0; j < cfg.n_tap; j++) {
      cs[i] = c;
      js[i] = j;
      if (calibrated) {
        us[i] = tapTbl[j] + dtcInj.delayCycles(c) + cfg.route_inj_cycles;
      } else {
        us[i] = j / cfg.n_tap + c / g;
      }
      i += 1;
    }
  }
  return { js, cs, us };
}

/** Produce injection command stream (computed-domain, pre-latency). */
export function runInjection(
  cfg: SimConfig,
  xNominal: ArrayLike<number>,
  rFb: ArrayLike<number>,
  dtcInj: DTCModel,
  tapTbl: ArrayLike<number>,
  ditherStream: Mulberry32 | null = null,
  dsmStream: Mulberry32 | null = null,
): InjectionResult {
  const g = configG(cfg);
  const n = xNominal.length;
  const tapStep = Math.floor(g / cfg.n_tap); // 32 LSB

  const uIdeal = uInjIdeal(xNominal, cfg.z0_cycles);

  // --- actuator mode 'dsm_only' (spec section 7.1): no fractional actuator
  if (cfg.actuator_mode === 'dsm_only') {
    const uInjAnalogConst = tapTbl[0] + dtcInj.delayCycles(0) + cfg.route_inj_cycles;
    const uInjAnalog = new Float64Array(n).fill(uInjAnalogConst);
    return {
      u_INJ_ideal: uIdeal,
      R_INJ: new Float64Array(n),
      j_INJ: new Float64Array(n),
      c_INJ: new Float64Array(n),
      u_INJ_digital: new Float64Array(n),
      u_INJ_analog: uInjAnalog,
    };
  }

  // --- R_INJ per architecture mode ---
  // ('qnc', spec section 7.2: always the modular reverse of the
  // cancellation-DTC code, regardless of arch_mode)
  const rInj = new Float64Array(n);
  if (cfg.arch_mode === 'D' || cfg.actuator_mode === 'qnc') {
    for (let k = 0; k < n; k++) {
      rInj[k] = pymod(cfg.r_zero - rFb[k], g);
    }
  } else {
    // A, B, C: independent quantizer instance on G * u_INJ_ideal
    const q = makeQuantizer(cfg.quantizer);
    // distinct deterministic DSM state per spec section 7 ('dsm_inj' stream)
    if (dsmStream !== null) {
      if (cfg.quantizer === 'ef1') {
        (q as ErrorFeedbackFirstOrder).seedState(dsmStream.next());
      } else if (cfg.quantizer === 'mash11') {
        const acc1 = dsmStream.next();
        const acc2 = dsmStream.next();
        (q as Mash11).seedState(acc1, acc2);
      } else if (cfg.quantizer === 'mash111') {
        const acc1 = dsmStream.next();
        const acc2 = dsmStream.next();
        const acc3 = dsmStream.next();
        (q as Mash111).seedState(acc1, acc2, acc3);
      }
    }
    const useDither = cfg.dither_amp_lsb > 0.0 && ditherStream !== null;
    for (let k = 0; k < n; k++) {
      let u = g * uIdeal[k];
      if (useDither && ditherStream !== null) {
        u += cfg.dither_amp_lsb * (ditherStream.next() + ditherStream.next() - 1.0);
      }
      rInj[k] = pymod(q.quantize(u), g);
    }
  }

  // --- tap + DTC decode ---
  const jInj = new Float64Array(n);
  const cInj = new Float64Array(n);
  if (cfg.inj_mapping === 'naive') {
    for (let k = 0; k < n; k++) {
      jInj[k] = Math.floor(rInj[k] / tapStep);
      cInj[k] = pymod(rInj[k], tapStep);
    }
  } else {
    const calibrated = cfg.inj_mapping === 'calibrated';
    const { js, cs, us } = candidates(cfg, dtcInj, tapTbl, calibrated);
    for (let k = 0; k < n; k++) {
      const target = rInj[k] / g;
      // first minimum -> smaller c, then smaller j (c-major enumeration)
      let best = 0;
      let bestErr = Math.abs(wrapCycles(target - us[0]));
      for (let i = 1; i < us.length; i++) {
        const err = Math.abs(wrapCycles(target - us[i]));
        if (err < bestErr) {
          bestErr = err;
          best = i;
        }
      }
      jInj[k] = js[best];
      cInj[k] = cs[best];
    }
  }

  const uInjDigital = new Float64Array(n);
  const uInjAnalog = new Float64Array(n);
  for (let k = 0; k < n; k++) {
    uInjDigital[k] = pymod(tapStep * jInj[k] + cInj[k], g) / g;
    uInjAnalog[k] = tapTbl[jInj[k]] + dtcInj.delayCycles(cInj[k]) + cfg.route_inj_cycles;
  }

  return {
    u_INJ_ideal: uIdeal,
    R_INJ: rInj,
    j_INJ: jInj,
    c_INJ: cInj,
    u_INJ_digital: uInjDigital,
    u_INJ_analog: uInjAnalog,
  };
}

/**
 * Pairwise reverse error (digital or analog layer, spec section 7):
 *
 * e_pair = wrapCycles(u_FB_actual + u_INJ_actual - u_zero_offset)
 * u_zero_offset = R_zero / G
 */
export function ePair(
  uFb: ArrayLike<number>,
  uInj: ArrayLike<number>,
  rZero: number,
  g: number,
): Float64Array {
  const tmp = new Float64Array(uFb.length);
  for (let k = 0; k < uFb.length; k++) {
    tmp[k] = uFb[k] + uInj[k] - rZero / g;
  }
  return wrapCyclesArr(tmp);
}
